New remix tree link placement option

DNin01 opened the pull request:
Resolves #3676

### Changes

Moved the location of the remix tree button. It is now located next to the remixes side pane as a link.

![The remixes list with a "Remix tree" link](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/4e4f4320-ff0c-4c2f-9023-6e136c19827a)

It is also possible to change this behavior back to what it was before.

### Reason for changes

The previous placement of the remix tree button uses up more space than I think is necessary, and how often does an average user visit the remix tree, anyway?

### Tests

Tested in Edge 115 to the best of my ability, including after my refactor. Dynamically enabling, disabling and changing settings and switching between the project page and the editor worked great!

DNin01 committed:
Change default remix tree link placement

DNin01 committed:
Format code

DNin01 committed:
Say link instead of button

WorldLanguages commented on addons/remix-tree-button/addon.json:
```diff
@@ -21,7 +25,30 @@
+  "versionAdded": "1.2.0",
+  "latestUpdate": {
+    "version": "1.33.0",
+    "temporaryNotice": "The link to the remix tree has moved. If you'd like to go back to the old behavior, you can change the setting below.",
```
Not sure if it's a good idea to force this... I bet many users won't find the button

WorldLanguages replied:
Unless we do something like this: https://github.com/ScratchAddons/ScratchAddons/pull/3603#issuecomment-961820896

WorldLanguages replied:
We could also try changing the default but only for users who have this addon disabled at the moment of the update (and for new users too, of course).

DNin01 replied:
> Not sure if it's a good idea to force this... I bet many users won't find the button
>
> We could also try changing the default but only for users who have this addon disabled at the moment of the update (and for new users too, of course).

I was definitely thinking about that.

DNin01 replied:
Turns out, I accidentally set the default to "new" instead of "replace", so the default hasn't changed, but I'm currently reconsidering what the default should be anyway.

DNin01 replied:
The default is now "new".

Secret-chest commented:
The old remixes link, though.

DNin01 commented:
> The old remixes link, though.

Which one are you taking about?

DNin01 commented:
Hang on, I forgot about l10n.

I'm working on another feature for this addon so give me a minute...

DNin01 committed:
Add option to turn icon into link

Secret-chest commented:
Wasn't there a "remixes" link in the remixes column?

mxmou commented:
> > The old remixes link, though.
> 
> Which one are you taking about?

Probably the "View all" link which this change removes.

DNin01 commented:
> Wasn't there a "remixes" link in the remixes column?

The one that goes to `/remixes`? If the user enables this addon, then they most likely aren't asking for that page.

DNin01 committed:
Add some comments to userscript

DNin01 committed:
Edit some comments

DNin01 committed:
Change default link placement back to original

Secret-chest commented:
> The one that goes to /remixes? If the user enables this addon, then they most likely aren't asking for that page.

They're both useful, in different ways.

DNin01 committed:
Use style.removeProperty

DNin01 committed:
Option to open in new tab

DNin01 committed:
Format code

DNin01 committed:
Fix redirection happening multiple times

DNin01 committed:
Format code

DNin01 committed:
Merge branch 'ScratchAddons:master' into move-remix-tree-link

Samq64 commented on addons/remix-tree-button/addon.json:
```diff
@@ -21,7 +25,42 @@
+      "default": "new"
+    },
+    {
+      "name": "Open remix tree by clicking remixes icon next to @heart.svg@star.svg",
```
I think this should be one of the options above instead of its own toggle

DNin01 replied:
This would be a good opportunity to allow multiple of these to be turned on at the same time. Maybe I should split everything into toggles so you don't have to lose one button to add another.

Samq64 commented on addons/remix-tree-button/main.js:
```diff
@@ -1,35 +1,174 @@
+    }
+  }
+
+  function undoLoadRemixButton() {
+    // Removes the button.
+    if (document.querySelector("#scratchAddonsRemixTreeBtn")) {
+      document.querySelector("#scratchAddonsRemixTreeBtn").style.display = "none";
+    }
+  }
```
Could [addon.tab.displayNoneWhileDisabled](https://scratchaddons.com/docs/reference/addon-api/addon.tab/#using-addontabdisplaynonewhiledisabled-dynamicdisable) be used here? It would also mean you wouldn't need to change the display type when the addon is dynamically re-enabled.

DNin01 replied:
I'll look into it. It might be more complicated when dealing with settings that can be dynamically changed.

Samq64 replied:
Oh, right, I guess that's the problem with having multiple placements for the same addon.

Samq64 commented on addons/remix-tree-button/main.js:
```diff
@@ -1,35 +1,174 @@
+    if (addon.settings.get("open-with-icon") && !addon.self.disabled) {
+      if (addon.settings.get("new-tab") || e.ctrlKey || e.metaKey) {
+        window.open(
+          `https://scratch.mit.edu/projects/${window.location.href.split("projects")[1].split("/")[1]}/remixtree`,
```
Just adding `remixtree` would be enough assuming there's a trailing slash in the URL, but if you want to be sure it would probably be better to store the project ID in a variable.

Samq64 commented on addons/remix-tree-button/main.js:
```diff
@@ -1,35 +1,174 @@
+  // Initialize:
+  initialSetup();
+
+  function initialSetup() {
```
I don't think this needs to be a function.

DNin01 committed:
Let's try this again

Starting over, this time focusing on clean code. Removing option to click remix icon on statistics (at least for now).

DNin01 commented:
Just redid everything and fixed a bug or two. It's so much easier to look at now.

I also removed the setting to click the 🌀 icon to go to the remix tree. It seems like you'd expect that to remix the project or something? Maybe I'm wrong. I might add it back, but I think we should prioritize reviewing what I've done already.

DNin01 committed:
Remove unused string

DNin01 committed:
Optimizations

DNin01 committed:
A couple refinements

DNin01 committed:
Shorten event listener functions

WorldLanguages commented:
Well, it's been a while... It's nice to see you've cleaned up the code with time though

WorldLanguages commented on addons/remix-tree-button/addon.json:
```diff
@@ -21,7 +25,36 @@
+  "latestUpdate": {
+    "version": "1.33.0",
+    "temporaryNotice": "The remix tree link can now be moved so it doesn't get in the way.",
+    "newSettings": ["type", "open-with-icon"]
```
Wrong setting listed

WorldLanguages commented on addons/remix-tree-button/main.js:
```diff
@@ -1,35 +1,121 @@
+          const linkSpan = link.firstChild;
+
+          if (!originalTextContent) {
+            originalTextContent = linkSpan.textContent; // Store what the button's text content was before the first time we change it (because different languages).
```
If you want, you can remove the global variable and just use `addon.tab.scratchMessage("project.viewAllInList")`

WorldLanguages commented on addons/remix-tree-button/addon.json:
```diff
@@ -21,7 +25,36 @@
-  "versionAdded": "1.2.0"
+  "versionAdded": "1.2.0",
+  "latestUpdate": {
+    "version": "1.33.0",
```
```suggestion
    "version": "1.36.0",
```

DNin01 committed:
Use addon.tab.scratchMessage

DNin01 committed:
Update latestUpdate properties

DNin01 commented:
> > The one that goes to /remixes? If the user enables this addon, then they most likely aren't asking for that page.
> 
> They're both useful, in different ways.

Maybe having it go to the full remixes list instead of the remix tree when Shift-clicking the link would be okay?

Secret-chest commented:
> Maybe having it go to the full remixes list instead of the remix tree when Shift-clicking the link would be okay?

Pretty undiscoverable. Why not add both?

DNin01 commented:
> > Maybe having it go to the full remixes list instead of the remix tree when Shift-clicking the link would be okay?
> 
> Pretty undiscoverable. Why not add both?

I could add it to the addon description or in a note or in a `title` property or something.

DNin01 commented:
A Shift alternative wouldn't be touch-screen friendly, however.

BroJac5246 commented:
> > > The one that goes to /remixes? If the user enables this addon, then they most likely aren't asking for that page.
> > 
> > They're both useful, in different ways.
> 
> Maybe having it go to the full remixes list instead of the remix tree when Shift-clicking the link would be okay?

Isn't that the browser shortcut for open page in new window?

DNin01 commented:
> > Maybe having it go to the full remixes list instead of the remix tree when Shift-clicking the link would be okay?
> 
> Isn't that the browser shortcut for open page in new window?

Never mind, then.

WorldLanguages commented:
Well, the default behavior of the addon is not changed, so it should be fine. We could also think of creative ways to keep both the "View all" and remix tree buttons, maybe by showing a tree icon instead of "remix tree", but I'm not sure if all languages have space for that. In touchscreen it would also be difficult to touch a small tree icon.

Secret-chest commented:
Just make it like in 2.0 where you can click the remix icon to view a tree.

DNin01 commented:
> Just make it like in 2.0 where you can click the remix icon to view a tree.

And make it possible for it to be enabled and for the two other options to be disabled?

That's actually an interesting idea.

DNin01 committed:
Merge branch 'ScratchAddons:master' into move-remix-tree-link
